File: PYTHON_WordSearch/tabuleiro.py
# -*- Mode: Python3; coding: utf-8; indent-tabs-mpythoode: nil; tab-width: 4 -*-

'''
*****************************************
*****   Python - LibreOffice 6.2    *****
*****   adaptado de tabuleiro.bas   *****
*****************************************
'''

from random import randint

# Constantes
CELULAVAZIA = "-"
PORCENTAGEM = 0.5

# Palavra
class Palavra(object):

	def __init__(self, texto):
		self.txt = texto
		self.msg = "falhou"
		self.posX = 0
		self.posY = 0
		self.posH = 0
		self.posV = 0
		self.uso  = 0
		
# Tabuleiro
class Tabuleiro(object):

	def __init__(self):
		self.lista = []
		self.matriz = ""
		self.tamanho = 0
		self.numPalavras = 0
		self.inseridos   = 0

	# Converte posição de vetor em coordenada de matriz
	def coordenada(self, posicao):

		posY = int(posicao - 1) / self.tamanho
		posX = posicao - self.tamanho * posY

		return [posY + 1, posX]
	
	# Converte coordenada da matriz em posição de vetor
	def posicao(self, x, y):

		return x + self.tamanho * (y - 1)

	# Validar coordenada na matriz
	def validar(self, texto, x, y, h, v):

		tamanho = self.tamanho - 1
		caracteres = len(texto) - 1

		# limites do tabuleiro
		if (x <= 0 or y <= 0):
			return False
		if (x > tamanho or y > tamanho):
			return False
		if (x + caracteres > tamanho and h == 1):
			return False
		if (y + caracteres > tamanho and v == 1):
			return False

		return True

	# Posicionar na matriz
	def dispor(self, offset, texto, x, y, h, v):

		retorno = self.validar(texto, x, y, h, v)
		if (not retorno):
			return retorno
		for i in range(0, len(texto)):
			pos = self.posicao(x + i * h, y + i * v)
			if (offset[pos] != CELULAVAZIA and offset[pos] != texto[i]):
				return False
			offset = offset[:pos - 1] + texto[i] + offset[pos:]

		# armazenar
		self.matriz = offset

		return retorno

	# Tabuleiro modo texto
	def tabuleiro_txt(self):

		resposta = ""
		for i in range(0, len(self.matriz)):
			if (i % self.tamanho == 0 and i != 0):
				resposta += "\n"				
			resposta += self.matriz[i]					

		return resposta

	# Informacao sobre posicoes das palavras
	def resumo(self):

		resposta = []
		for p in self.lista:
			s = p.txt + " (" + str(p.posX) + "," + str(p.posY) + "): " + p.msg
			resposta += [s]

		return resposta
	
# Inserir palavras no tabuleiro
def tabular(lista):

	# inicializar tabuleiro
	T = Tabuleiro()
	T.numPalavras = len(lista)

	if (T.numPalavras == 0):
		# retornar tabuleiro vazio
		return T

	for i in range(0, T.numPalavras):
		
		# excluir dados inválidos
		if (lista[i] == ""): continue	
		try:
			# somente letras
			lista[i].isalpha()
		except Exception as e:
			continue

		# maior palavra
		if (T.tamanho < len(lista[i])):
			T.tamanho = len(lista[i])

		# preparar palavra
		p = lista[i].upper()
		T.lista += [Palavra(p)]

	# aumentar tamanho com constante porcentagem
	T.tamanho = T.tamanho + int(T.tamanho * PORCENTAGEM)
	
	# aumentar tamanho se maior palavra for menor que numero de palavras
	if (T.tamanho < T.numPalavras):
		T.tamanho = T.numPalavras

	# inicializar matriz
	T.matriz = (T.tamanho * T.tamanho) * CELULAVAZIA

	# inserir
	for p in T.lista:
		offset = T.matriz

		# posição aleatória
		x = randint(1, T.tamanho - 1)
		y = randint(1, T.tamanho - 1)

		# tentativa na diagonal para baixo
		if (not p.uso):
			if (T.dispor(offset, p.txt, x, y, 1, 1)):
				p.posH = 1
				p.posV = 1
				p.msg = "Diagonal para baixo"
				p.uso = True

		# tentativa na horizontal para a direita
		if (not p.uso):
			if (T.dispor(offset, p.txt, x, y, 1, 0)):
				p.posH = 1
				p.posV = 0
				p.msg = "Horizontal para direita"
				p.uso = True

		# tentativa na vertical para baixo
		if (not p.uso):
			if (T.dispor(offset, p.txt, x, y, 0, 1)):
				p.posH = 0
				p.posV = 1
				p.msg = "Vertical para baixo"
				p.uso = True

		if (p.uso):
			p.posX = x 
			p.posY = y
			T.inseridos += 1

	# retornar tabuleiro
	return T

def teste():

	conjuntoPalavras = [
		# dados válidos
		"amor",
		"carro",
		"jabuticaba",
		"lua",
		"martelo",
		"palhaço",
		"pão",
		"peixe",
		"pera",
		"raio",
		"sapo",
		"sol",
		"vida",
		# dados inválidos
		1.0,
		-10,
		[10]
	]

	T = tabular(conjuntoPalavras)

	print("Tabuleiro:")
	print(T.tabuleiro_txt())
	print(T.tamanho * CELULAVAZIA)
	print("Palavras:", T.numPalavras)
	print("Inseridos:", T.inseridos)
	print(T.tamanho * CELULAVAZIA)

	for s in T.resumo():
		print(s)

if __name__ == '__main__':
	teste()

File: Kotlin_WordSearch/src/main/kotlin/Tabuleiro.kt
private const val PORCENTAGEM = 0.5F
private const val DELIMITADOR = ';'
private const val CELULA_VAZIA = "-"
private const val EOL = '\n'

class Tabuleiro {

    private var lista : MutableList<Palavra>    // palavras
    private var matriz: String                  // vetor tabuleiro
    private var tamanho: Int                    // matriz quadrada
    private var numPalavras : Int               // lista
    private var inseridos : Int                 // tabuleiro

    init {
        matriz = ""
        tamanho = 0
        numPalavras = 0
        inseridos = 0
        lista = ArrayList()
    }

    constructor(lista: String) {
        if (!gerarTabuleiro(lista.split(DELIMITADOR))) {
            println("Erro ao gerar o Tabuleiro!")
        }
    }

    constructor(lista: List<String>) {
        if (!gerarTabuleiro(lista)) {
            println("Erro ao gerar o Tabuleiro!")
        }
    }

    /* Tabuleiro */
    private fun gerarTabuleiro(lista: List<String>) : Boolean {

        // validar lista
        for (valor in lista) {
            if (valor.length == 1) {
                println("item '$valor' não é válido!")
                continue
            }
            if (!valor.all{it.isLetter()}) {
                println("item '$valor' não é válido!")
                continue
            }
            if (valor.isNotEmpty() && valor.isNotBlank()) {
                // adicionar palavra a lista principal
                this.lista.add(Palavra(texto = valor.lowercase()))
                // atualizar tamanho com maior palavra
                if (tamanho < valor.length) {
                    tamanho = valor.length
                }
            }
        }

        // atualizar numero de palavras
        numPalavras = this.lista.size
        if (numPalavras == 0) {
            // nada a fazer
            return false
        }

        // aumentar tamanho se maior palavra for menor que número de palavras
        if (tamanho < numPalavras) {
            tamanho = numPalavras
        }

        // aumentar tamanho com constante porcentagem
        tamanho += (tamanho * PORCENTAGEM).toInt()

        // inicializar matriz - string
        matriz = CELULA_VAZIA.repeat(tamanho * tamanho)

        // inserir
        for (palavra in this.lista) {
            // posição aleatória
            val x = (0 until tamanho - 1).random()
            val y = (0 until tamanho - 1).random()

            // tentativa 1
            if (!palavra.ativo) {
                if (dispor(palavra.texto, x, y, 1, 1)) {
                    palavra.msg = "Diagonal para baixo"
                    palavra.ativo = true
                    palavra.posH = 1
                    palavra.posV = 1
                }
            }

            // tentativa 2
            if (!palavra.ativo) {
                if (dispor(palavra.texto, x, y, 1, 0)) {
                    palavra.msg = "Horizontal para direita"
                    palavra.ativo = true
                    palavra.posH = 1
                    palavra.posV = 0
                }
            }

            // tentativa 3
            if (!palavra.ativo) {
                if (dispor(palavra.texto, x, y, 0, 1)) {
                    palavra.msg = "Vertical para baixo"
                    palavra.ativo = true
                    palavra.posH = 0
                    palavra.posV = 1
                }
            }

            // atualizar
            if (palavra.ativo) {
                palavra.posX = x
                palavra.posY = y
                inseridos += 1
            }
        }

        // tabuleiro ok
        return true
    }

    /* Posicionar na matriz */
    private fun dispor(texto: String, x: Int, y: Int, h: Int, v: Int) : Boolean {

        if (!validar(texto, x, y, h, v)) {
            // posição inválida
            return false
        }

        var offset = matriz
        for (i in texto.indices) {
            val pos = posicao(x + i * h, y + i * v)
            if (offset[pos] != CELULA_VAZIA[0] && offset[pos] != texto[i]) {
                // ocupada com letra de outra palavra
                return false
            }
            // organizar
            var txt = ""
            for (j in offset.indices) {
                txt += if (j == pos) {
                    texto[i]
                } else {
                    offset[j]
                }
            }
            // atualizar
            offset = txt
        }

        // atualizar
        matriz = offset

        // inserção ok
        return true
    }

    /* Validar coordenada na matriz */
    private fun validar(texto: String, x: Int, y: Int, h: Int, v: Int) : Boolean {

        // limites do tabuleiro
        if (x < 0 || y < 0) {
            return false
        }
        if (x > tamanho || y > tamanho) {
            return false
        }
        if (x + texto.length > tamanho && h == 1) {
            return false
        }
        if (y + texto.length > tamanho && v == 1) {
            return false
        }

        return true
    }

    /* Converte coordenada da matriz em posição de vetor */
    private fun posicao(x: Int, y: Int) : Int = x + tamanho * y

    /* Word Search */
    private fun ofuscar() : String {
        var matriz = ""
        val letras = "abcdefghijklmnopqrstuvwxyz"
        for (caracter in this.matriz) {
            matriz += if (caracter == CELULA_VAZIA[0])
                letras[(letras.indices).random()]
            else
                caracter
        }

        return matriz
    }

    /* Tabuleiro modo texto */
    fun tabuleiro(respota: Boolean = false) : String {

        val matriz: String = if (respota) {
            this.matriz
        } else {
            this.ofuscar()
        }

        var saida = ""
        for (i in matriz.indices) {
            if (i % tamanho == 0 && i != 0) {
                saida += EOL
            }
            saida += matriz[i]
        }

        return saida
    }

    /* Informação sobre posições das palavras */
    fun resumo() : String {

        var resposta = "Respostas:$EOL"
        for (palavra in this.lista) {
            resposta += palavra.resumo() + EOL
        }
        resposta += """Palavras: $numPalavras ${EOL}Inseridas: $inseridos $EOL"""

        return resposta
    }
}
